server/backup_runner.go

```diff
@@ -51,13 +58,69 @@ func (s *BackupRunner) Run(ctx context.Context, wg *sync.WaitGroup) {
 					}
 				}()
 				s.startAutoBackups(ctx, runningTasks, &mu)
+				s.downloadBinlogFiles(ctx)
 			}()
 		case <-ctx.Done(): // if cancel() execute
+			s.downloadBinlogWg.Wait()
 			return
 		}
 	}
 }
 
+func (s *BackupRunner) downloadBinlogFiles(ctx context.Context) {
+	instanceList, err := s.server.store.FindInstanceWithDatabaseBackupEnabled(ctx, db.MySQL)
+	if err != nil {
+		log.Error("Failed to retrieve MySQL instance list with at least one database backup enabled", zap.Error(err))
+		return
+	}
+
+	for _, instance := range instanceList {
+		s.downloadBinlogMu.Lock()
```

Comment:
Typically, Lock() comes with defer Unlock(). We could lock the whole for-loop if that makes things easy.

```
Lock()
defer Unlock()
for {
  ...
}
```